pkg/controller/certificates/clustertrustbundlepublisher/publisher.go

```diff
@@ -0,0 +1,230 @@
+				Name: "ca_cert_publisher_cluster_trust_bundles",
+			},
+		),
+	}
+	p.ctbInformer = setupSignerNameFilteredCTBInformer(p.client, p.signerName)
+	p.ctbLister = certlisters.NewClusterTrustBundleLister(p.ctbInformer.GetIndexer())
+	p.ctbListerSynced = p.ctbInformer.HasSynced
+
+	_, err := p.ctbInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			p.queue.Add("")
+		},
+		UpdateFunc: func(_, _ interface{}) {
+			p.queue.Add("")
+		},
+		DeleteFunc: func(_ interface{}) {
+			p.queue.Add("")
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to register ClusterTrustBundle event handler: %w", err)
+	}
+	p.ca.AddListener(p.caContentChangedListener())
+
+	return p, nil
+}
+
+func (p *ClusterTrustBundlePublisher) caContentChangedListener() dynamiccertificates.Listener {
+	return caContentListener(func() {
+		p.queue.Add("")
+	})
+}
+
+func (p *ClusterTrustBundlePublisher) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+	defer p.queue.ShutDown()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting ClusterTrustBundle CA cert publisher controller")
+	defer logger.Info("Shutting down ClusterTrustBundle CA cert publisher controller")
+
+	go p.ctbInformer.Run(ctx.Done())
+
+	if !cache.WaitForNamedCacheSync("cluster trust bundle", ctx.Done(), p.ctbListerSynced) {
+		return
+	}
+
+	// init the signer syncer
+	p.queue.Add("")
+	go wait.UntilWithContext(ctx, p.runWorker(), time.Second)
+
+	<-ctx.Done()
+}
+
+func (p *ClusterTrustBundlePublisher) runWorker() func(context.Context) {
+	return func(ctx context.Context) {
+		for p.processNextWorkItem(ctx) {
+		}
+	}
+}
+
+// processNextWorkItem deals with one key off the queue. It returns false when
+// it's time to quit.
+func (p *ClusterTrustBundlePublisher) processNextWorkItem(ctx context.Context) bool {
+	key, quit := p.queue.Get()
+	if quit {
+		return false
+	}
+	defer p.queue.Done(key)
+
+	if err := p.syncClusterTrustBundle(ctx, key); err != nil {
+		utilruntime.HandleError(fmt.Errorf("syncing %q failed: %v", key, err))
+		p.queue.AddRateLimited(key)
+		return true
+	}
+
+	p.queue.Forget(key)
+	return true
+}
+
+func (p *ClusterTrustBundlePublisher) syncClusterTrustBundle(ctx context.Context, _ string) (err error) {
+	startTime := time.Now()
+	defer func() {
+		recordMetrics(startTime, err)
+		klog.FromContext(ctx).V(4).Info("Finished syncing ClusterTrustBundle", "signerName", p.signerName, "elapsedTime", time.Since(startTime))
+	}()
+
+	caBundle := p.ca.CurrentCABundleContent()
+	bundleName := bundleName(p.signerName, caBundle)
+
+	bundle, err := p.ctbLister.Get(bundleName)
+	if apierrors.IsNotFound(err) {
+		_, err = p.client.CertificatesV1alpha1().ClusterTrustBundles().Create(ctx, &certificatesv1alpha1.ClusterTrustBundle{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: bundleName,
+			},
+			Spec: certificatesv1alpha1.ClusterTrustBundleSpec{
+				SignerName:  p.signerName,
+				TrustBundle: string(p.ca.CurrentCABundleContent()),
```

Comment:
Keep a single meaning for `caBundle` during this sync loop call.

```suggestion
				TrustBundle: caBundle,
```